src/librustdoc/html/highlight.rs

```diff
@@ -727,27 +766,76 @@ fn string<T: Display>(
+                                    && matches!(klass, Class::Macro(_))
+                                {
+                                    Ok(generate_macro_def_id_path(href_context, *def_id))
+                                } else {
+                                    Err(e)
+                                }
+                            })
+                            .ok()
                     }
                     LinkFromSrc::Primitive(prim) => format::href_with_root_path(
                         PrimitiveType::primitive_locations(context.tcx())[prim],
                         context,
-                        Some(context_info.root_path),
+                        Some(href_context.root_path),
                     )
                     .ok()
                     .map(|(url, _, _)| url),
                 }
             })
         {
-            write!(out, "<a class=\"{}\" href=\"{}\">{}</a>", klass.as_html(), href, text_s);
-            return;
+            write!(out, "<a class=\"{}\" href=\"{}\">{}", klass.as_html(), href, text_s);
+            return Some("</a>");
         }
     }
-    write!(out, "<span class=\"{}\">{}</span>", klass.as_html(), text_s);
+    write!(out, "<span class=\"{}\">{}", klass.as_html(), text_s);
+    Some("</span>")
+}
+
+/// This function is to get the external macro path because they are not in the cache used n
+/// `href_with_root_path`.
+fn generate_macro_def_id_path(href_context: &HrefContext<'_, '_, '_>, def_id: DefId) -> String {
+    let tcx = href_context.context.shared.tcx;
+    let crate_name = tcx.crate_name(def_id.krate).to_string();
+    let cache = &href_context.context.cache();
+
+    let relative = tcx.def_path(def_id).data.into_iter().filter_map(|elem| {
+        // extern blocks have an empty name
+        let s = elem.data.to_string();
+        if !s.is_empty() { Some(s) } else { None }
+    });
+    // Check to see if it is a macro 2.0 or built-in macro
+    let mut path = if matches!(
+        CStore::from_tcx(tcx).load_macro_untracked(def_id, tcx.sess),
+        LoadedMacro::MacroDef(def, _)
+            if matches!(&def.kind, ast::ItemKind::MacroDef(ast_def)
+                if !ast_def.macro_rules)
+    ) {
+        once(crate_name.clone()).chain(relative).collect()
+    } else {
+        vec![crate_name.clone(), relative.last().expect("relative was empty")]
+    };
+
+    let url_parts = match cache.extern_locations[&def_id.krate] {
+        ExternalLocation::Remote(ref s) => vec![s.trim_end_matches('/')],
+        ExternalLocation::Local => vec![href_context.root_path.trim_end_matches('/'), &crate_name],
+        ExternalLocation::Unknown => panic!("unknown crate"),
+    };
+
+    let last = path.pop().unwrap();
+    let last = format!("macro.{}.html", last);
+    if path.is_empty() {
+        format!("{}/{}", url_parts.join("/"), last)
+    } else {
+        format!("{}/{}/{}", url_parts.join("/"), path.join("/"), last)
+    }
 }
```

Comment:
I had a bunch of comments on this function so I figured I'd just try rewriting it locally to my taste. Let me know what you think!

```suggestion
/// Return the URL for a macro at the given DefId.
/// 
/// This needs its own function because macros are not in the cache used by `href_with_root_path`.
fn generate_macro_def_id_path(href_context: &HrefContext<'_, '_, '_>, def_id: DefId) -> String {
    let tcx = href_context.context.shared.tcx;
    let relative_path = tcx.def_path(def_id).data.into_iter().filter_map(|elem| {
        // extern blocks have an empty name
        let s = elem.data.to_string();
        if !s.is_empty() { Some(s) } else { None }
    });

    let crate_name = tcx.crate_name(def_id.krate).to_string();
    // In macros 2.0, macros can have a path. If available, find that path.
    // https://github.com/rust-lang/rfcs/blob/master/text/1584-macros.md
    let mut path = match CStore::from_tcx(tcx).load_macro_untracked(def_id, tcx.sess) {
        LoadedMacro::MacroDef(def, _)
            if matches!(&def.kind,
                ast::ItemKind::MacroDef(ast_def) if !ast_def.macro_rules) => {
            once(crate_name.clone()).chain(relative_path).collect()
        },

        _ => vec![crate_name.clone(), relative_path.last().expect("macro's path has no name")],
    };

    let macro_name = path.pop().expect("macro has empty path when linkifying");
    let root = match href_context.context.cache().extern_locations[&def_id.krate] {
        ExternalLocation::Remote(remote) => remote.clone(),
        ExternalLocation::Local => format!("{}{}", href_context.root_path, crate_name),
        ExternalLocation::Unknown => panic!("unknown crate"),
    };
    format!("{}{}/macro.{}.html", root, path.join("/"), macro_name)
}
```

Reply:
It's actually a bit similar to what I wrote except mine is a bit longer but I find it easier to read. You will tell me what you think of it. :)